h2o-core/src/main/java/water/util/FrameSizeMonitor.java

```diff
@@ -0,0 +1,178 @@
+package water.util;
+
+import water.*;
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.parser.FVecParseWriter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class FrameSizeMonitor extends MrFun<FrameSizeMonitor> {
+
+    private static final String ENABLED_PROP = "util.frameSizeMonitor.enabled";
+    private static final String SAFE_COEF_PROP = "util.frameSizeMonitor.safetyCoefficient";
+    private static final String SAFE_FREE_MEM_DEFAULT_COEF = "0.1";
+
+    private static final boolean ENABLED;
+    private static final float SAFE_FREE_MEM_COEF;
+    private static final int SLEEP_MS = 100;
+    private static final int MB = 1024 * 1024;
+    
+    static {
+        ENABLED = H2O.getSysBoolProperty(ENABLED_PROP, false);
+        SAFE_FREE_MEM_COEF = Float.parseFloat(H2O.getSysProperty(SAFE_COEF_PROP, SAFE_FREE_MEM_DEFAULT_COEF));
+    }
+
+    private final Key<Job> jobKey;
+    private final Map<FVecParseWriter, Long> writers = new HashMap<>();
+    private final long totalMemory = getTotalMemory();
+
+    public FrameSizeMonitor(Key<Job> jobKey) {
+        this.jobKey = jobKey;
+    }
+
+    public static FrameSizeMonitor start(Key<Job> jobKey) {
+        FrameSizeMonitor task = new FrameSizeMonitor(jobKey);
+        if (ENABLED) {
+            H2O.submitTask(new LocalMR(task, 1));
+        }
+        return task;
```

Comment:
maybe it would be better if this returned null in case the feature is not enabled, the rest of the code is ready for that